scipy/cluster/hierarchy.py

```diff
@@ -1395,6 +1400,184 @@ def cut_tree(Z, n_clusters=None, height=None):
     return groups.T
 
 
+def cut_tree_balanced(Z, max_cluster_size):
+    """
+    Given a linkage matrix Z and max cluster size, return a balanced cut tree.
+
+    The function looks recursively along the hierarchical tree, from the root
+    (single cluster gathering all the samples) to the leaves (i.e. the clusters
+    with only one sample), retrieving the biggest possible clusters containing
+    a number of samples lower than a given maximum. If a cluster at a specific
+    tree level contains a number of samples higher than the given maximum, it
+    is ignored and its offspring (smaller) sub-clusters are taken into
+    consideration. If the cluster contains a number of samples lower than the
+    given maximum, it is taken as result and its offspring sub-clusters not
+    further processed.
+
+    Parameters
+    ----------
+    Z : ndarray
+        The linkage matrix resulting from calling `ward` or `linkage`. I.e. it 
+        contains the hierarchical clustering encoded as a linkage matrix.
+    max_cluster_size : int
```

Comment:
Thank you very much for this sensible comment. Indeed it would be nice to follow the same approach as in `cut_tree` and perform multiple cuts in one go. However, in the case of `cut_tree_balanced` the code might get too complex. The problem ist that for each iteration (representing the current tree level) we mark the vector positions at lower tree levels as deleted for further processing if the current one has already `<= max_cluster_size` elements. In other words, we alter the underlying data structure `full_cut`, so that no further values of `max_cluster_size` could be processed using the same data structure. Thus, it might get complex and expensive to maintain multiple copies of `full_cut` to enable that functionality. Please let me know if this is clear, and whether you see any other way of overcoming this limitation.

Reply:
Right, you'd have to just do this in an outer loop over different `max_cluster_size`s, and it would not be more efficient than the user doing that same loop - only more convenient, and consistent in API with `cut_tree`.

But let's hold off on this one for now, because it may make the second return value ugly.

Reply:
I agree, let's hold off on this one for now.